builtin/providers/softlayer/resource_softlayer_ssh_key.go

```diff
@@ -0,0 +1,125 @@
+package softlayer
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	datatypes "github.com/maximilien/softlayer-go/data_types"
+)
+
+func resourceSoftLayerSSHKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSoftLayerSSHKeyCreate,
+		Read:   resourceSoftLayerSSHKeyRead,
+		Update: resourceSoftLayerSSHKeyUpdate,
+		Delete: resourceSoftLayerSSHKeyDelete,
+
+		Schema: map[string]*schema.Schema{
+			"id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"public_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"fingerprint": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceSoftLayerSSHKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client).sshKeyService
+
+	// Build up our creation options
+	opts := datatypes.SoftLayer_Security_Ssh_Key{
+		Label: d.Get("name").(string),
+		Key: d.Get("public_key").(string),
+	}
+
+	res, err := client.CreateObject(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating SSH Key: %s", err)
+	}
+
+	d.SetId(strconv.Itoa(res.Id))
+	log.Printf("[INFO] SSH Key: %d", res.Id)
+
+	return resourceSoftLayerSSHKeyRead(d, meta)
+}
+
+func resourceSoftLayerSSHKeyRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client).sshKeyService
+
+	keyId, _ := strconv.Atoi(d.Id())
+
+	key, err := client.GetObject(keyId)
+	if err != nil {
+		// If the key is somehow already destroyed, mark as
+		// succesfully gone
+		if strings.Contains(err.Error(), "404 Not Found") {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving SSH key: %s", err)
+	}
+
+	d.Set("id", key.Id)
+	d.Set("name", key.Label)
+	d.Set("public_key", strings.TrimSpace(key.Key))
+	d.Set("fingerprint", key.Fingerprint)
+
+	return nil
+}
+
+func resourceSoftLayerSSHKeyUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client).sshKeyService
+
+	keyId, _ := strconv.Atoi(d.Id())
+
+	key, err := client.GetObject(keyId)
```

Comment:
Usually the convention within a resource implementation is for ``Update`` to construct a fresh object from the ``schema.ResourceData``, rather than fetching it from the backing API. Note that ``Read`` is usually called earlier in the flow unless the user intentionally skips it, so you can generally assume that the state is up-to-date with the remote resource by the time we get here, as long as the ``Read`` implementation is good.

Trusting that the ``ResourceData`` contains the correct values means that the remote resource should 100% match the Terraform configuration once this function returns, whereas in your current implementation there is a (admittedly very small) chance that the remote object could drift after the plan is created and only the ``name`` attribute is guaranteed to match at the end as a result.